src/abstractdataframe/join.jl

```diff
@@ -81,16 +81,622 @@ end
+            else
+                for (left_i, right_i) in Iterators.product(left_cur:left_new - 1,
+                                                           right_cur:right_new - 1)
+                    push!(left_ixs, left_i)
+                    push!(right_ixs, right_i)
+                end
+            end
+            left_cur, left_val = left_new, left_tmp
+            left_new, left_tmp = find_next_range(left, left_cur, left_val)
+            right_cur, right_val = right_new, right_tmp
+            right_new, right_tmp = find_next_range(right, right_cur, right_val)
+        elseif isless(left_val, right_val)
+            left_cur, left_val = left_new, left_tmp
+            left_new, left_tmp = find_next_range(left, left_cur, left_val)
+        else
+            right_cur, right_val = right_new, right_tmp
+            right_new, right_tmp = find_next_range(right, right_cur, right_val)
+        end
+    end
+
+    return left_ixs, right_ixs
+end
+
+# optimistically assume that shorter table does not have duplicates in on column
+# if this is not the case we call _innerjoin_dup
+# which efficiently uses the work already done and continues with the more
+# memory expensive algorithm that allows for duplicates
+function _innerjoin_unsorted(left::AbstractArray, right::AbstractArray{T}) where {T}
+    dict = Dict{T, Int}()
+
+    for (idx_r, val_r) in enumerate(right)
+        # we use dict_index to make sure the following two operations are fast:
+        # - if index is found - fall back to algorithm allowing duplicates
+        # - if index is not found - add it
+        dict_index = Base.ht_keyindex2!(dict, val_r)
+        dict_index > 0 && return _innerjoin_dup(left, right, dict, idx_r)
+        Base._setindex!(dict, idx_r, val_r, -dict_index)
+    end
+
+    left_ixs = Int[]
+    right_ixs = Int[]
+
+    for (idx_l, val_l) in enumerate(left)
+        # we use dict_index to make sure the following two operations are fast:
+        # - if index is found - get it and process it
+        # - if index is not found - do nothing
+        dict_index = Base.ht_keyindex(dict, val_l)
+        if dict_index > 0 # -1 if key not found
+            @inbounds idx_r = dict.vals[dict_index]
+            push!(left_ixs, idx_l)
+            push!(right_ixs, idx_r)
+        end
+    end
+    return left_ixs, right_ixs
+end
+
+extrema_missing(x::AbstractVector{Missing}) = 1, 0
+
+function extrema_missing(x::AbstractVector{T}) where {T<:Union{Integer, Missing}}
+    try
+        return extrema(skipmissing(x))
+    catch
+        S = nonmissingtype(T)
+        return S(1), S(0)
+    end
+end
+
+function _innerjoin_unsorted_int(left::AbstractVector{<:Union{Integer, Missing}},
+                                 right::AbstractVector{<:Union{Integer, Missing}})
+    minv, maxv = extrema_missing(right)
+
+    if (maxv - minv) > 128 && (maxv - minv) ÷ 2 > length(right) &&
+       (minv < typemin(Int) + 2 || maxv > typemax(Int) - 3)
```

Comment:
Is this supposed to use `&&`? If `(maxv - minv) ÷ 2 > length(right)` we always want to use the fallback, right? Also you need to ensure that `maxv - minv` and `Int(maxv) - Int(minv) + 2` don't overflow. I found it simpler to use `big` everywhere since it's cheap.

Reply:
> we always want to use the fallback, right?

I thought that if the pool is small (up to 128) we never want to use the fallback. Allocating 128 element vector is super cheap and then later it is much faster to operate on vector rather than dict.

but the following `&&` should be `||` indeed. I will fix it.

> don't overflow

You are right - I will fix it.